tidy3d/updater.py

```diff
@@ -1,49 +1,120 @@
-"""Class holding methods to upgrade old versions of :class:`.Simulation` json dictionaries."""
-from typing import Dict
+"""Utilities for converting between tidy3d versions."""
+import json
+import functools
+
+import pydantic as pd
 
 from .version import __version__
+from .log import FileError, SetupError
+
+
+"""Storing version numbers."""
+
+
+class Version(pd.BaseModel):
+    """Stores a version number (excluding patch)."""
+
+    major: int
+    minor: int
+
+    @classmethod
+    def from_string(cls, string) -> "Version":
+        """Return Version from a version string."""
+        try:
+            major, minor, _ = string.split(".")
+            version = cls(major=major, minor=minor)
+        except Exception as e:
+            raise SetupError(f"version string {version_string} can't be parsed.") from e
+        return version
+
+    @property
+    def as_tuple(self):
+        """version as a tuple, leave out patch for now."""
+        return (self.major, self.minor, None)
+
+    def __hash__(self):
+        """define a hash."""
+        return hash(self.as_tuple)
+
+
+CurrentVersion = Version.from_string(__version__)
+
+"""Class for updating simulation objects."""
+
+
+class Updater(pd.BaseModel):
+    """Converts a tidy3d simulation.json file to an up-to-date Simulation instance."""
+
+    sim_dict: dict
+
+    @classmethod
+    def from_file(cls, fname: str) -> "Updater":
+        """Dictionary representing the simulation loaded from file."""
+
+        try:
+            with open(fname, "r") as f:
+                sim_dict = json.load(f)
+        except Exception as e:
+            raise FileError(f"Could not load file {fname}") from e
+        return cls(sim_dict=sim_dict)
+
+    @property
+    def version(self) -> Version:
+        """Version of the supplied file."""
+        version_string = self.sim_dict.get("version")
+        if version_string is None:
+            raise SetupError("Could not find a version in the supplied json.")
+        return Version.from_string(version_string)
+
+    def update_to_current(self) -> dict:
+        """Update supplied simulation dictionary to current version."""
+        while self.version != CurrentVersion:
+            update_fn = UPDATE_MAP.get(self.version)
+            if update_fn is None:
+                raise SetupError(f"version {self.version} not found in update map.")
+            self.sim_dict = update_fn(self.sim_dict)
+        self.sim_dict['version'] == __version__
+        return self.sim_dict
+
+
+"""Update conversion functions."""
+
+# versions will be dynamically mapped in this table when the update functions are initialized.
+UPDATE_MAP = {}
```

Comment:
I think this may be wrong, `UPDATE_MAP` may still be global only within this module. Just unsure about this.

Reply:
`UPDATE_MAP` will only be referenced in the `Updater` class. It's not meant to be imported anywhere, nor do we plan to import `UPDATE_MAP` from anywhere else, otherwise we'd probably have a problem.  The idea is to have these updater functions added to the UPDATE_MAP at import time, but maybe the other approach is to do it like you had before where each of the functions are static methods and the UPDATE_MAP is some kind of class variable?

Reply:
we can make the name more unique.

Reply:
I think this is fine now that it's more clear to me how these things work.